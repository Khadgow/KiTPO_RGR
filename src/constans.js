export const Routes = {
  REGISTER: '/register',
  ROOT: '/',
  LOGIN: '/login',
  CREATE: '/create',
  INFO: '/info/:id',
  USERS: '/users',
  PROFILE: '/profile',
  UPDATE_CURRENT_USER_PROFILE: '/profile/update',
  UPDATE_USER_PROFILE: '/profile/update/:id',
  MULTI_STEP_FORM: '/multi-step-form',
}

export const Api = {
  USER: {
    LOGIN: '/login',
    REGISTER: '/register',
    LOAD_USER: '/user',
    GET_ALL_USERS: '/users',
    GET_USER: (id) => `/users/${id}`,
    UPDATE_USER: (id) => `/users/${id}`,
    UPDATE_USER_AVATAR: (id) => `/users/${id}/avatar`,
    GET_USER_LANGUAGES: (id) => `/users/${id}/languages`,
    ADD_USER_LANGUAGE: (id) => `/users/${id}/languages`,
    UPDATE_USER_LANGUAGE: ({ userId, languageCode }) =>
      `/users/${userId}/languages/${languageCode}`,
    DELETE_USER_LANGUAGE: ({ userId, languageCode }) =>
      `/users/${userId}/languages/${languageCode}`,
  },
  POSTS: {
    CREATE: '/posts',
    GET_ALL_POSTS: '/posts',
    GET_POST: (id) => `/posts/${id}`,
    EDIT_POST: (id) => `/posts/${id}`,
    DELETE_POST: (id) => `/posts/${id}`,
    LIKE_POST: (id) => `/posts/${id}/like`,
  },
  EDUCATION: {
    CREATE: (userId) => `/users/${userId}/educations`,
    GET_ALL_EDUCATIONS: (userId) => `/users/${userId}/educations`,
    UPDATE_EDUCATION: ({ userId, educationId }) => `/users/${userId}/educations/${educationId}`,
    DELETE_EDUCATION: ({ userId, educationId }) => `/users/${userId}/educations/${educationId}`,
  },
  EXPERIENCE: {
    CREATE: (userId) => `/users/${userId}/experiences`,
    GET_ALL_EXPERIENCES: (userId) => `/users/${userId}/experiences`,
    UPDATE_EXPERIENCE: ({ userId, experienceId }) => `/users/${userId}/experiences/${experienceId}`,
    DELETE_EXPERIENCE: ({ userId, experienceId }) => `/users/${userId}/experiences/${experienceId}`,
  },
  LANGUAGE: {
    CREATE: '/languages',
    GET_ALL_LANGUAGES: '/languages',
    UPDATE_LANGUAGE: (languageId) => `/languages/${languageId}`,
    DELETE_LANGUAGE: (languageId) => `/languages/${languageId}`,
  },
}

export const Roles = {
  user: 'user',
  admin: 'admin',
}

export const FormsValidation = {
  email: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
}

export const ApiTokenStorageKey = 'userToken'

export const dateFormat = 'yyyy-MM-dd'
