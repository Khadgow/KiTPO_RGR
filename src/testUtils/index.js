export * from './jestUtils'
export * from './serverUtils'
