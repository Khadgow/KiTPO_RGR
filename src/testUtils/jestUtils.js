import React, { useEffect } from 'react'
import { render } from '@testing-library/react'
import { Provider, ReactReduxContext, useDispatch, useSelector } from 'react-redux'
import { ConnectedRouter, routerMiddleware } from 'connected-react-router'
import { configureStore } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
import { createBrowserHistory } from 'history'

import { actions } from 'ducks/user/slice'
import userSelectors from 'ducks/user/selectors'
import { Spinner } from 'components/Spinner/Spinner'
import rootReducer from 'store/root-reducer'
import rootSaga from 'store/root-saga'

let store
let history

beforeEach(() => {
  history = createBrowserHistory()
  const sagaMiddleware = createSagaMiddleware()

  store = configureStore({
    reducer: rootReducer(history),
    middleware: [sagaMiddleware, routerMiddleware(history)],
  })

  sagaMiddleware.run(rootSaga)
})

const AllTheProviders = ({ children }) => {
  return (
    <Provider store={store} context={ReactReduxContext}>
      <ConnectedRouter history={history}>{children}</ConnectedRouter>
    </Provider>
  )
}

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

const UserLoadingContainer = ({ children }) => {
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(actions.loadUserRequest())
  }, [dispatch])
  const user = useSelector(userSelectors.selectUser)
  const isAuthenticated = useSelector(userSelectors.selectAuthState)
  if (!isAuthenticated || !user) {
    return <Spinner />
  }

  return children
}

const UserLoadingWrapper = ({ children }) => (
  <AllTheProviders>
    <UserLoadingContainer>{children}</UserLoadingContainer>
  </AllTheProviders>
)

export const renderWithUserLoading = (ui, options) =>
  render(ui, { wrapper: UserLoadingWrapper, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
