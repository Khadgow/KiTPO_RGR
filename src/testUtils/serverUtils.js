import { setupServer } from 'msw/node'
import { rest } from 'msw'

import { generateApiPath } from '../utils/generateApiPath'
import { Api } from '../constans'

export const servers = {
  contactInformationForm: ({ mockedUser, requestFunc }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.put(generateApiPath(Api.USER.UPDATE_USER(mockedUser.id)), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  educationInformationForm: ({ mockedUser, mockedEducations, requestFunc }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.get(
        generateApiPath(Api.EDUCATION.GET_ALL_EDUCATIONS(mockedUser.id)),
        (req, res, ctx) => {
          return res(
            ctx.status(200),
            ctx.json({
              data: { educations: mockedEducations },
            }),
          )
        },
      ),
      rest.put(
        generateApiPath(Api.EDUCATION.UPDATE_EDUCATION({ userId: mockedUser.id, educationId: 2 })),
        async (req, res, ctx) => {
          const body = await req.json()
          requestFunc(body)
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
      rest.delete(
        generateApiPath(Api.EDUCATION.DELETE_EDUCATION({ userId: mockedUser.id, educationId: 2 })),
        async (req, res, ctx) => {
          requestFunc()
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
      rest.post(generateApiPath(Api.EDUCATION.CREATE(mockedUser.id)), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  languagesForm: ({
    mockedUser,
    mockedLanguages,
    mockedUserLanguages,
    putRequestFunc,
    postRequestFunc,
  }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.get(generateApiPath(Api.LANGUAGE.GET_ALL_LANGUAGES), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { languages: mockedLanguages },
          }),
        )
      }),
      rest.get(generateApiPath(Api.USER.GET_USER_LANGUAGES(mockedUser.id)), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { languages: mockedUserLanguages },
          }),
        )
      }),
      rest.put(
        generateApiPath(
          Api.USER.UPDATE_USER_LANGUAGE({ userId: mockedUser.id, languageCode: 'en' }),
        ),
        async (req, res, ctx) => {
          const body = await req.json()
          putRequestFunc({ level: body.level })
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
      rest.post(
        generateApiPath(Api.USER.ADD_USER_LANGUAGE(mockedUser.id)),
        async (req, res, ctx) => {
          const body = await req.json()
          postRequestFunc({
            code: body.code,
            level: body.level,
          })
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
    ),
  workExperienceForm: ({ mockedUser, mockedWorkExperiences, requestFunc }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.get(
        generateApiPath(Api.EXPERIENCE.GET_ALL_EXPERIENCES(mockedUser.id)),
        (req, res, ctx) => {
          return res(
            ctx.status(200),
            ctx.json({
              data: { experiences: mockedWorkExperiences },
            }),
          )
        },
      ),
      rest.put(
        generateApiPath(
          Api.EXPERIENCE.UPDATE_EXPERIENCE({ userId: mockedUser.id, experienceId: 2 }),
        ),
        async (req, res, ctx) => {
          const body = await req.json()
          requestFunc(body)
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
      rest.delete(
        generateApiPath(
          Api.EXPERIENCE.DELETE_EXPERIENCE({ userId: mockedUser.id, experienceId: 2 }),
        ),
        async (req, res, ctx) => {
          requestFunc()
          return res(
            ctx.status(200),
            ctx.json({
              success: true,
            }),
          )
        },
      ),
      rest.post(generateApiPath(Api.EXPERIENCE.CREATE(mockedUser.id)), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  navbar: ({ mockedUser }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
    ),
  createPostForm: ({ requestFunc }) =>
    setupServer(
      rest.post(generateApiPath(Api.POSTS.CREATE), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  editPostForm: ({ requestFunc, mockedPost }) =>
    setupServer(
      rest.get(generateApiPath(Api.POSTS.GET_POST(1)), async (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { post: mockedPost },
          }),
        )
      }),
      rest.put(generateApiPath(Api.POSTS.EDIT_POST(1)), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
      rest.delete(generateApiPath(Api.POSTS.DELETE_POST(1)), async (req, res, ctx) => {
        requestFunc({ id: mockedPost.id })
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  postListContainer: ({ mockedPosts }) =>
    setupServer(
      rest.get(generateApiPath(Api.POSTS.GET_ALL_POSTS), async (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { posts: mockedPosts, meta: { last_page: 1 } },
          }),
        )
      }),
    ),
  profileContainer: ({ mockedUser }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
    ),
  updateProfileForm: ({ mockedUser, requestFunc }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.LOAD_USER), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.get(generateApiPath(Api.USER.GET_USER(mockedUser.id)), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { user: mockedUser },
          }),
        )
      }),
      rest.put(generateApiPath(Api.USER.UPDATE_USER(mockedUser.id)), async (req, res, ctx) => {
        const body = await req.json()
        requestFunc(body)
        return res(
          ctx.status(200),
          ctx.json({
            success: true,
          }),
        )
      }),
    ),
  usersListContainer: ({ mockedUsers }) =>
    setupServer(
      rest.get(generateApiPath(Api.USER.GET_ALL_USERS), (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            data: { users: mockedUsers },
          }),
        )
      }),
    ),
}
