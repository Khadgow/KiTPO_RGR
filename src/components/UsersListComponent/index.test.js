import React from 'react'

import { render } from 'testUtils'
import { Roles } from 'constans'

import { UsersListComponent } from './index'

const mockedUsers = [
  {
    id: 1,
    name: 'name',
    lastName: 'last name',
    email: 'some@mail.com',
    role: Roles.user,
    birthdate: '2000-01-01',
  },
  {
    id: 2,
    name: 'name 2',
    lastName: 'last name 2',
    email: 'some2@mail.com',
    role: Roles.admin,
    birthdate: '1990-01-01',
  },
]

describe('UsersListComponent', () => {
  test('Snapshot test', async () => {
    const { asFragment } = render(<UsersListComponent users={mockedUsers} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
