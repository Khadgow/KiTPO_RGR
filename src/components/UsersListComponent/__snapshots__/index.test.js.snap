// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersListComponent Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="usersList"
  >
    <div>
      <p>
        formFields.name: name
      </p>
      <p>
        formFields.lastName: last name
      </p>
      <p>
        formFields.email: some@mail.com
      </p>
      <p>
        formFields.role: user
      </p>
      <p>
        formFields.birthdate: 2000-01-01
      </p>
      <button
        type="button"
      >
        Update
      </button>
      <hr />
    </div>
    <div>
      <p>
        formFields.name: name 2
      </p>
      <p>
        formFields.lastName: last name 2
      </p>
      <p>
        formFields.email: some2@mail.com
      </p>
      <p>
        formFields.role: admin
      </p>
      <p>
        formFields.birthdate: 1990-01-01
      </p>
      <button
        type="button"
      >
        Update
      </button>
      <hr />
    </div>
  </div>
</DocumentFragment>
`;
